core/rawdb/freezer_table.go

```diff
@@ -853,6 +853,33 @@ func (t *freezerTable) sizeNolock() (uint64, error) {
 	return total, nil
 }
 
+// sizeHidden returns the total data size of hidden items in the freezer table.
+func (t *freezerTable) sizeHidden() (uint64, error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	hiddenItems := t.itemHidden.Load()
+	offsetItems := t.itemOffset.Load()
+	if hiddenItems <= offsetItems {
+		// Return 0 if there is no hidden items
+		return 0, nil
+	}
+
+	indices, err := t.getIndices(offsetItems, hiddenItems-offsetItems)
```

Comment:
You don't really need to resolve all the hidden items and calculate the total hidden data size.
Instead, you can just retrieve the offset corresponds to the first *non-hidden* item, and the
offset is actually the hidden data size.

```go
// sizeHidden returns the total data size of hidden items in the freezer table.
func (t *freezerTable) sizeHidden() (uint64, error) {
	t.lock.RLock()
	defer t.lock.RUnlock()

	hidden, offset := t.itemHidden.Load(), t.itemOffset.Load()
	if hidden <= offset {
		return 0, nil
	}
	indices, err := t.getIndices(hidden, 1)
	if err != nil {
		return 0, err
	}
	return uint64(indices[0].offset), nil
}
```

Reply:
> Thank you for the advice! I will try this method.
> 
> It seems my test cases failed on Windows environments in AppVeyor builds. The files were not dropped when there were only hidden items after truncation. Is it by design or a bug? It might fail this method.

The truncateTail failed in the first place with `rename C:\Users\appveyor\AppData\Local\Temp\1\336933468 C:\Users\appveyor\AppData\Local\Temp\1\truncate-tail-12670623878883326278.ridx: Access is denied.` [https://ci.appveyor.com/project/karloku/go-ethereum/builds/47842081/job/tlckukoxk9mi3j7a#L969](https://ci.appveyor.com/project/karloku/go-ethereum/builds/47842081/job/tlckukoxk9mi3j7a#L969)

I will try to fix it.